pkg/proxy/proxy.go

```diff
@@ -228,67 +229,100 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	ctx, cancel := context.WithTimeout(context.Background(), p.Timeout)
 	defer cancel()
 
 	// Serialize commands sent to a specific VIN to avoid some complexities associated with sharing
 	// the vehicle.Vehicle object. VCSEC commands fail if they arrive out of order, anyway.
 	if err := p.lockVIN(ctx, vin); err != nil {
 		writeJSONError(w, http.StatusServiceUnavailable, err)
-		return
+		return err
 	}
 	defer p.unlockVIN(vin)
 
-	log.Debug("Executing %s on %s", command, vin)
-	if req.Method != http.MethodPost {
-		writeJSONError(w, http.StatusMethodNotAllowed, nil)
-		return
-	}
-	car, err := acct.GetVehicle(ctx, vin, p.commandKey, p.sessions)
-	if err != nil || car == nil {
-		writeJSONError(w, http.StatusInternalServerError, err)
-		return
+	car, commandToExecuteFunc, err := p.loadVehicleAndCommandFromRequest(ctx, acct, w, req, command, vin)
+	if err != nil {
+		return err
 	}
 
 	if err := car.Connect(ctx); err != nil {
 		writeJSONError(w, http.StatusInternalServerError, err)
-		return
+		return err
 	}
 	defer car.Disconnect()
 
 	if err := car.StartSession(ctx, nil); err == protocol.ErrProtocolNotSupported {
 		p.markUnsupportedVIN(vin)
 		p.forwardRequest(acct.Host, w, req)
-		return
+		return err
 	} else if err != nil {
 		writeJSONError(w, http.StatusInternalServerError, err)
-		return
+		return err
 	}
 	defer car.UpdateCachedSessions(p.sessions)
 
-	err = execute(ctx, req, car, command)
-	if err == ErrUseRESTAPI {
-		p.forwardRequest(acct.Host, w, req)
-		return
+	if err = commandToExecuteFunc(car); err == vehicle.ErrCommandUseRESTAPI {
+		return err
 	}
 	if protocol.IsNominalError(err) {
 		writeJSONError(w, http.StatusOK, err)
-		return
+		return err
 	}
 	if err != nil {
 		writeJSONError(w, http.StatusInternalServerError, err)
-		return
+		return err
 	}
 
 	w.Header().Add("Content-Type", "application/json")
 	fmt.Fprintln(w, "{\"response\":{\"result\":true,\"reason\":\"\"}}")
+	return nil
+}
+
+func (p *Proxy) loadVehicleAndCommandFromRequest(ctx context.Context, acct *account.Account, w http.ResponseWriter, req *http.Request,
+	command, vin string) (*vehicle.Vehicle, func(*vehicle.Vehicle) error, error) {
+
+	log.Debug("Executing %s on %s", command, vin)
+	if req.Method != http.MethodPost {
+		writeJSONError(w, http.StatusMethodNotAllowed, nil)
+		return nil, nil, fmt.Errorf("Wrong http method")
+	}
+
+	commandToExecuteFunc, err := extractCommandAction(ctx, req, command)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	car, err := acct.GetVehicle(ctx, vin, p.commandKey, p.sessions)
+	if err != nil || car == nil {
+		writeJSONError(w, http.StatusInternalServerError, err)
+		return nil, nil, err
+	}
+
+	return car, commandToExecuteFunc, err
+}
+
+func extractCommandAction(ctx context.Context, req *http.Request, command string) (func(*vehicle.Vehicle) error, error) {
+	var params vehicle.RequestParameters
+	body, err := io.ReadAll(req.Body)
```

Comment:
does `req.Body` limit the input size? If not, then we should use a LimitReader here to prevent DoS.

Reply:
I was concerned about clients DoSing the proxy server, not the proxy server DoSing Tesla's backend. But it looks like the Go library buts some sane limits in place (https://go.dev/src/net/http/request.go#L1234).